### PR TITLE
feat(Button): Alternative (secondary) modes and consistent height [GM-126]

### PR DESCRIPTION
**Pull request by codecaaron**
### Overview

These button updates are being used in both [Alerts](https://github.com/Codecademy/client-modules/pull/1412) and [Banners](https://github.com/Codecademy/client-modules/pull/1412).

- Adds a new prop `variant?: 'primary' | 'secondary'`
- Button font-size is always 700.
- Button heights are now consistent between the 4 types of UI buttons (and normal=42 and small=32 ) for consistency when buttons are adjacent.
- IconButton is now fully square.
- New colors: `yellow-400` | `blue-800`

Other changes
- Migrates some shared styles to variant API, and updated call signature.
- Updates some inconsistent but functionally equivalent rules (background -> background-color) across button types.
- Adds storybook grid for seeing all button variants.

### Screenshots

Screen Shot 2021-02-17 at 8 54 18 AM
Screen Shot 2021-02-17 at 8 54 26 AM

### PR Checklist

- [x] Related to designs: [Figma](https://www.figma.com/file/orSca8PFzTB7qiuDw3QR8R/%5BWIP%5D-Button-Color-Variants?node-id=268%3A0)
- [x] Related to JIRA ticket: GM-126
- [x] I have run this code to verify it works
- [ ] This PR includes unit tests for the code change

### CODE REVIEW

File: packages/gamut/src/Button/shared.ts [hunk 1]
```diff
@@ -1,38 +1,59 @@
...
-export type ButtonProps = HTMLProps<HTMLLinkElement> &
-  HTMLProps<HTMLButtonElement> & {
-    mode?: 'dark' | 'light';
-  };
-
-export type ButtonSize = 'normal' | 'small';
+export type ButtonProps = Omit<
+  HTMLProps<HTMLLinkElement> & HTMLProps<HTMLButtonElement>,
+  'size'
+> & {
```
Comment (codecaaron) [inline packages/gamut/src/Button/shared.ts:7]: This is needed as the "size" key resolves to undefined in the intersection downstream.

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-illustrations@0.4.42-alpha.fbd27e.0<br/>@codecademy/gamut-labs@10.0.8-alpha.fbd27e.0<br/>@codecademy/gamut-styles@8.0.2-alpha.fbd27e.0<br/>@codecademy/gamut-tests@2.0.2-alpha.fbd27e.0<br/>@codecademy/gamut@23.0.2-alpha.fbd27e.0<br/>@codecademy/markdown-overrides@0.5.2-alpha.fbd27e.0<br/>@codecademy/styleguide@27.0.2-alpha.fbd27e.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://602d1ad5eafe690bc8343e16--gamut-preview.netlify.app>https://602d1ad5eafe690bc8343e16--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/602d1ad5eafe690bc8343e16>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-illustrations@0.4.42-alpha.792edf.0<br/>@codecademy/gamut-labs@10.0.8-alpha.792edf.0<br/>@codecademy/gamut-styles@8.0.2-alpha.792edf.0<br/>@codecademy/gamut-tests@2.0.2-alpha.792edf.0<br/>@codecademy/gamut@23.0.2-alpha.792edf.0<br/>@codecademy/markdown-overrides@0.5.2-alpha.792edf.0<br/>@codecademy/styleguide@27.0.2-alpha.792edf.0<br/></pre>

File: packages/gamut/src/Button/ButtonInner.tsx [hunk 2]
```diff
@@ -8,11 +8,34 @@ export const ButtonInner = styled('span', {
...
+
+    > * {
+      vertical-align: middle;
+
+      &:first-of-type {
+        margin-left: -1px;
+      }
+      &:last-of-type {
+        margin-right: -1px;
+      }
```
Comment (codecaaron) [inline packages/gamut/src/Button/ButtonInner.tsx:38]: This is probably unnecessary if we keep this component flex, but I suspect there may be times where an inline-block container will make arbitrary children easier to style without having to override the button.

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-illustrations@0.4.42-alpha.f367be.0<br/>@codecademy/gamut-labs@10.0.8-alpha.f367be.0<br/>@codecademy/gamut-styles@8.0.2-alpha.f367be.0<br/>@codecademy/gamut-tests@2.0.2-alpha.f367be.0<br/>@codecademy/gamut@23.0.2-alpha.f367be.0<br/>@codecademy/markdown-overrides@0.5.2-alpha.f367be.0<br/>@codecademy/styleguide@27.0.2-alpha.f367be.0<br/></pre>

File: packages/gamut/src/Button/ButtonInner.tsx [hunk 3]
```diff
@@ -8,11 +8,34 @@ export const ButtonInner = styled('span', {
...
-    ${fontSmoothing()}
+    &:before,
+    &:after {
+      content: '';
+      display: inline-block;
+      width: 1px;
+      height: 100%;
+      margin-left: -1px;
+      vertical-align: middle;
+    }
```
Comment (codecaaron) [inline packages/gamut/src/Button/ButtonInner.tsx:28]: This is a trick to make the line height however much space is in the div vertically and allow the rest of the content to align itself to the center using vertical align-middle.

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://602d222e707bc614a044604c--gamut-preview.netlify.app>https://602d222e707bc614a044604c--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/602d222e707bc614a044604c>Deploy Logs</a></p>

**Comment by jakemhiller** [comment]
@codecaaron what do you think about keeping "mode" as just `light|dark` and adding something else to switch the color? i feel like having a simple `mode` switch as a pattern will be nice if we ever do any sort of automated light/dark mode. Also wondering, if this new color is `${mode}-alt`, what will the next color option be?

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-illustrations@0.4.42-alpha.d90ae6.0<br/>@codecademy/gamut-labs@10.0.8-alpha.d90ae6.0<br/>@codecademy/gamut-styles@8.0.2-alpha.d90ae6.0<br/>@codecademy/gamut-tests@2.0.2-alpha.d90ae6.0<br/>@codecademy/gamut@23.0.2-alpha.d90ae6.0<br/>@codecademy/markdown-overrides@0.5.2-alpha.d90ae6.0<br/>@codecademy/styleguide@27.0.2-alpha.d90ae6.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://602d2fef5fe38823e663c548--gamut-preview.netlify.app>https://602d2fef5fe38823e663c548--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/602d2fef5fe38823e663c548>Deploy Logs</a></p>

**Review by dreamwasp — APPROVED** [review]
🙌

File: packages/gamut/src/Button/ButtonInner.tsx [hunk 4]
```diff
@@ -8,11 +8,35 @@ export const ButtonInner = styled('span', {
...
+    font-weight: ${fontWeight.title};
+    text-align: center;
     transition: ${timing.fast} background-color, ${timing.fast} box-shadow,
       ${timing.fast} color;
 
-    ${fontSmoothing()}
+    &:before,
+    &:after {
+      content: '';
+      display: inline-block;
```
Comment (JoshuaKGoldberg) [inline packages/gamut/src/Button/ButtonInner.tsx:23]: Heh, no love for `inline-flex`?

File: packages/styleguide/stories/Atoms/Button/index.stories.mdx [hunk 5]
```diff
@@ -90,3 +92,27 @@ All but CTA button support a size of either `normal` (default) or `small`.
 </Canvas>
 
 <Props story="IconButton" />
+
+## Button Variants
+
+We have 2 button color modes: `light` and `dark` which are meant to be used in their respective contexts. For differentiation we also provide an alternate color scheme for lower priority buttons when used adjacently. These are `light-alt` and `dark-alt` respectively.
```
Comment (JoshuaKGoldberg) [inline packages/styleguide/stories/Atoms/Button/index.stories.mdx:98]:
```suggestion
We have 2 button color modes: `light` and `dark`.
For differentiation we also provide an alternate color scheme for lower priority buttons when used alongside other buttons: `light-alt` and `dark-alt`.
```

File: packages/gamut/src/Button/shared.ts [hunk 6]
```diff
@@ -1,38 +1,59 @@
...
+  mode?: 'dark' | 'light' | 'dark-alt' | 'light-alt';
```
Comment (JoshuaKGoldberg) [inline packages/gamut/src/Button/shared.ts:8]: I'm not sure how I feel about `-alt` as the naming convention. In my mind "alt" refers to a bit more different of a visual, like with colors... but I can't think of a better one. 🤷

File: yarn.lock [hunk 7]
```diff
@@ -7374,7 +7374,7 @@ error-stack-parser@^2.0.6:
   dependencies:
     stackframe "^1.1.1"
 
-es-abstract@^1.17.0-next.0, es-abstract@^1.17.2, es-abstract@^1.17.4:
+es-abstract@^1.17.0-next.0, es-abstract@^1.17.0-next.1, es-abstract@^1.17.2, es-abstract@^1.17.4:
```
Comment (JoshuaKGoldberg) [inline yarn.lock:7377]: Unrelated?

**Review by JoshuaKGoldberg — APPROVED** [review]
Nice!

File: packages/styleguide/stories/Atoms/Button/index.stories.mdx [hunk 8]
```diff
@@ -90,3 +92,27 @@ All but CTA button support a size of either `normal` (default) or `small`.
...
+- Use `light` and `dark` modes for your primary buttons.
+- Use `light-alt` and `dark-alt` modes for secondary buttons.
+
+### Light Mode
+
+<Canvas>
+  <Story name="Light Mode">
+    <ButtonScale mode="light" />
+  </Story>
+</Canvas>
```
Comment (dreamwasp) [inline packages/styleguide/stories/Atoms/Button/index.stories.mdx:110]: I think an additional code example (like 
```
<TextButton
  href=""
  mode="light-alt"
  size="normal"
>
  Click Me
</TextButton>
``` 
here would be helpful

**Review by dreamwasp — COMMENTED** [review]
There's a couple weird things with the story: the new modes aren't showing up under the 'mode' dropdown of the existing buttons + something Weird is happening with the IconButton story
image 
but I think is unrelated to this PR.

Other than those small things ✨

File: yarn.lock (hunk 7 above)
Comment (codecaaron) [reply yarn.lock:7377]: Yeah for whatever reason this came up after a lerna clean => yarn

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-illustrations@0.4.42-alpha.5b9a3c.0<br/>@codecademy/gamut-labs@10.0.8-alpha.5b9a3c.0<br/>@codecademy/gamut-styles@8.0.2-alpha.5b9a3c.0<br/>@codecademy/gamut-tests@2.0.2-alpha.5b9a3c.0<br/>@codecademy/gamut@23.0.2-alpha.5b9a3c.0<br/>@codecademy/markdown-overrides@0.5.2-alpha.5b9a3c.0<br/>@codecademy/styleguide@27.0.2-alpha.5b9a3c.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://602d4b180d211f0949f104c2--gamut-preview.netlify.app>https://602d4b180d211f0949f104c2--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/602d4b180d211f0949f104c2>Deploy Logs</a></p>

**Comment by codecaaron** [comment]
> @codecaaron what do you think about keeping "mode" as just `light|dark` and adding something else to switch the color? i feel like having a simple `mode` switch as a pattern will be nice if we ever do any sort of automated light/dark mode. Also wondering, if this new color is `${mode}-alt`, what will the next color option be?

Makes sense to me.  I switched this to be 2 dimensions `mode: 'dark' | 'light'` and `variant: 'primary' | 'secondary'`.  

I think the follow-up to that would be to make `mode` a reserved prop going forward and follow a similar structure here, and have a single pattern to update when we are able to automate it.

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-illustrations@0.4.42-alpha.185707.0<br/>@codecademy/gamut-labs@10.0.8-alpha.185707.0<br/>@codecademy/gamut-styles@8.0.2-alpha.185707.0<br/>@codecademy/gamut-tests@2.0.2-alpha.185707.0<br/>@codecademy/gamut@23.0.2-alpha.185707.0<br/>@codecademy/markdown-overrides@0.5.2-alpha.185707.0<br/>@codecademy/styleguide@27.0.2-alpha.185707.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://602d4ca037aefc0a3451413a--gamut-preview.netlify.app>https://602d4ca037aefc0a3451413a--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/602d4ca037aefc0a3451413a>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-illustrations@0.4.43-alpha.b00ee6.0<br/>@codecademy/gamut-labs@10.0.10-alpha.b00ee6.0<br/>@codecademy/gamut-styles@8.0.2-alpha.b00ee6.0<br/>@codecademy/gamut-tests@2.0.2-alpha.b00ee6.0<br/>@codecademy/gamut@23.1.1-alpha.b00ee6.0<br/>@codecademy/markdown-overrides@0.5.2-alpha.b00ee6.0<br/>@codecademy/styleguide@27.1.2-alpha.b00ee6.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://602d4f03bf1ccc1cba57259d--gamut-preview.netlify.app>https://602d4f03bf1ccc1cba57259d--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/602d4f03bf1ccc1cba57259d>Deploy Logs</a></p>
